RAG_supporters/dataset/label_calculator.py

```diff
@@ -0,0 +1,383 @@
+    
+    Args:
+        method: Normalization method
+        **kwargs: Additional arguments (e.g., temperature for softmax)
+        
+    Returns:
+        LabelNormalizer instance
+    """
+    if method == LabelNormalizationMethod.SOFTMAX:
+        temperature = kwargs.get('temperature', 1.0)
+        return SoftmaxNormalizer(temperature=temperature)
+    elif method == LabelNormalizationMethod.SOFTMAX_TEMPERATURE:
+        temperature = kwargs.get('temperature', 1.0)
+        if temperature == 1.0:
+            logger.warning("Using SOFTMAX_TEMPERATURE with default temperature=1.0")
+        return SoftmaxNormalizer(temperature=temperature)
+    elif method == LabelNormalizationMethod.L1:
+        return L1Normalizer()
+    elif method == LabelNormalizationMethod.L2:
+        return L2Normalizer()
+    else:
+        raise ValueError(f"Unknown normalization method: {method}")
+
+
+class LabelCalculator:
+    """
+    Calculates cluster membership labels for dataset samples.
+    
+    Supports 3 label types:
+    1. Source/question labels (from CSV or computed)
+    2. Steering embedding labels (computed from centroids)
+    3. Combined labels (for masking augmentation)
+    """
+    
+    def __init__(
+        self,
+        clustering_data: ClusteringData,
+        normalization_method: LabelNormalizationMethod = LabelNormalizationMethod.SOFTMAX,
+        temperature: float = 1.0
+    ):
+        """
+        Initialize label calculator.
+        
+        Args:
+            clustering_data: ClusteringData with centroids
+            normalization_method: Method to normalize distances to probabilities
+            temperature: Temperature for softmax normalization
+        """
+        self.clustering_data = clustering_data
+        self.normalizer = get_normalizer(
+            normalization_method, 
+            temperature=temperature
+        )
+        
+        if clustering_data.centroids is None:
+            raise ValueError("ClusteringData must have centroids for label calculation")
+        
+        self.n_clusters = clustering_data.n_clusters
+        logger.info(f"Initialized LabelCalculator with {self.n_clusters} clusters")
+        logger.info(f"Normalization: {normalization_method.value}")
+    
+    def calculate_source_labels_from_suggestions(
+        self,
+        suggestions: List[Dict],
+        suggestion_embeddings: Dict[str, np.ndarray]
+    ) -> np.ndarray:
+        """
+        Calculate source labels from suggestion texts and their embeddings.
+        
+        Computes average distance of all suggestion embeddings to each centroid,
+        then normalizes to probability distribution.
+        
+        Args:
+            suggestions: List of suggestion dicts with 'term' field
+            suggestion_embeddings: Dict mapping suggestion terms to embeddings
+            
+        Returns:
+            Probability vector of shape (n_clusters,)
+        """
+        if not suggestions:
```

Comment:
Silently returning uniform distribution can hide bugs. Consider raising an error or making it explicit via a parameter.